### DISCUSSION TITLE
Change Events for Open/Close of Secondary Menus

### DISCUSSION THREAD

**Discussion by MichaelWDenney** [kirill-grouchnikov/radiance#486] [Q&A] [answered]
Is there any means, such as a change event, to know when a secondary menu has been opened or closed?  I have logic that is updating a secondary menu so want it to pause while the menu is open and resume when the menu is closed.

**Comment by kirill-grouchnikov** [comment]
What do you want to pause? The opening and populating of the menu, or that process that is updating the menu content when it's not open?

**Reply by MichaelWDenney** [reply]
I want to pause the process updating the menu content.  The update can happen at any time and frequently.  If it happens while the menu is open, it causes entries in the menu to become invalid such that selecting them may have no result.

**Comment by kirill-grouchnikov** [comment, marked as answer]
Is `BaseCommand.SecondaryLifecycle` a good fit for your needs?

**Reply by MichaelWDenney** [reply]
I'll take a look.

**Reply by MichaelWDenney** [reply]
Unless I misunderstand, I think this could be used to pause the opening and populating, which is not what I need.

**Reply by kirill-grouchnikov** [reply]
It's not used to control the menu opening and closing, it is notifying the listener about the lifecycle of a popup menu.

**Reply by MichaelWDenney** [reply]
I spoke too soon.  So, maybe overriding onBeforeActivateSecondary and onAfterDeactivateSecondary could give me what I need?

**Reply by MichaelWDenney** [reply]
That actually seems to be working quite well.  Thanks!

By the way, if my process removes all entries from the menu, the "v" arrow on the bottom of the button does not get hidden until I force a refresh by moving my cursor over it.  Same thing happens in reverse when the processing causes an empty menu to get populated.  Is that a bug or should I do something in the code to force a refresh?

**Reply by kirill-grouchnikov** [reply]
That sounds like a bug. Ideally, it would have a small, self contained and compileable code that shows the issue, with the simplest setup that you can show for how the menu content is getting updated.

**Reply by MichaelWDenney** [reply]
Yeah, unfortunately that's not really the case.  But, at its core it is just calling `Command..getSecondaryContentModel().removeAllCommandGroups()` followed by repeated calls to `Command..getSecondaryContentModel().addCommandGroup()`.

I'll try to find time to set up something small in the demo code to reproduce the behaviour.

**Reply by MichaelWDenney** [reply]
Any reason using the `BaseCommand.SecondaryLifecycle` wouldn't work in the top-level application menu?  It works in submenus of the application menu, but when I do something similar in the top-level, `onBeforeActivateSecondary `and `onAfterDeactivateSecondary` don't appear to trigger.  Could it be because, in the application menu, the secondary menus don't so much activate/deactivate as a single secondary menu gets repopulated?

**Reply by kirill-grouchnikov** [reply]
Correct, in the ribbon application menu, the second-level items are displayed in a fixed panel that is not activated / deactivated as the primary menu items are activated.

**Reply by MichaelWDenney** [reply]
So, any way to trigger on the population of the panel with a particular "sub menu" set of items?  It's less critical as the case I was dealing with above, but it could be useful.

**Reply by kirill-grouchnikov** [reply]
Mmm, I guess it can be wired in that specific configuration to the same lifecycle events.

**Reply by MichaelWDenney** [reply]
So, from a coding perspective it looks like just another submenu...or at least closer to?  Yeah, that would be a nice way to handle it.

**Reply by kirill-grouchnikov** [reply]
I'll need to think about it, since all the methods in that interface get a `JPopupPanel`, but the secondary ribbon application menu does not have its own popup panel. Passing `null` feels a bit of an awkward workaround. Maybe something with a more generic way of specifying the type of container that hosts that content.

**Reply by MichaelWDenney** [reply]
Sure.  As I said, it's not critical to me.  At least not yet.  ;)
